ccl/backupccl: add memory monitor to external SST iterators in restore

rhu713 opened the pull request:
Previously, there was no limit on the amount of memory that can be used while
constructing edternal SST iterators during restore. This patch adds a memory
monitor to limit the amount of memory that can be used to construct external
SST iterators. If a restore processor fails to acquire enough memory to open
the next file for a restore span, it will send the iterator for all of the open
files it has accumulated so far, and wait until it can acquire the memory to
resume constructing the iterator for the remaining files.

The memory limit can be controlled via the new cluster setting
`bulkio.restore.per_processor_memory_limit`. Regardless of the setting,
however, the amount of memory used will not exceed
`COCKROACH_RESTORE_MEM_FRACTION` * `max SQL memory.` The new environment
variable `COCKROACH_RESTORE_MEM_FRACTION` defaults to 0.5.

Benchmarking using 10 iterators of 100 files each, each file is 24MiB in size.
```
./dev bench ./pkg/ccl/backupccl/  --filter BenchmarkIteratorMemory/fileCount=100$/iterCount=10$/rows=200000$/enc=false  --bench-mem -- --test_env=COCKROACH_GCS_SST_DIR=gs://rui-backup-test/bench --test_env=COCKROACH_S3_SST_DIR=s3://rui-crl/bench (plus credentials environment variables...)

GCS unencrypted: 5.0 MiB/file  
  5246162544 B/op

GCS encrypted: 8.9 MiB/file
  9404394496 B/op

S3 unencrypted: 1.4 MiB/file 
  1511477488 B/op

S3 encrypted: 1.7 MiB/file
  1834325232 B/op
```

Fixes: #102722 

Release note: None

Commit:
ccl/backupccl: add memory monitor to external SST iterators in restore

Previously, there was no limit on the amount of memory that can be used while
constructing edternal SST iterators during restore. This patch adds a memory
monitor to limit the amount of memory that can be used to construct external
SST iterators. If a restore processor fails to acquire enough memory to open
the next file for a restore span, it will send the iterator for all of the open
files it has accumulated so far, and wait until it can acquire the memory to
resume constructing the iterator for the remaining files.

The memory limit can be controlled via the new cluster setting
bulkio.restore.per_processor_memory_limit. Regardless of the setting,
however, the amount of memory used will not exceed
COCKROACH_RESTORE_MEM_FRACTION * max SQL memory. The new environment
variable COCKROACH_RESTORE_MEM_FRACTION defaults to 0.5.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93324)
<!-- Reviewable:end -->

benbardin commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -317,21 +365,68 @@ func (rd *restoreDataProcessor) openSSTs(
+		file := entry.Files[idx]
 		log.VEventf(ctx, 2, "import file %s which starts at %s", file.Path, entry.Span.Key)
 
+		if err := rd.mem.Grow(ctx, bytesPerFile); err != nil {
```
Separately, how would you feel about un-indenting some of this, i.e. `if err :=...; err == nil { continue}` Think that might be more readable? or would the `err == nil` be too surprising?

rhu713 replied:
I think this is okay for now to keep `err != nil` as the conditional. We have large err blocks elsewhere too.

benbardin commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -317,21 +365,68 @@ func (rd *restoreDataProcessor) openSSTs(
 		dirs = append(dirs, dir)
 		storeFiles = append(storeFiles, storageccl.StoreFile{Store: dir, FilePath: file.Path})
+		idx++
 		// TODO(pbardea): When memory monitoring is added, send the currently
```
TODO can go away, right?

shermanCRL commented:
What are next steps here?

rhu713 commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -317,21 +365,68 @@ func (rd *restoreDataProcessor) openSSTs(
+				storeFiles = make([]storageccl.StoreFile, 0, len(entry.Files)-idx)
+			}
+
+			time.Sleep(sstReaderMemWaitSec * time.Second)
```
yeah you are right, I'll fix this

rhu713 commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -317,21 +365,68 @@ func (rd *restoreDataProcessor) openSSTs(
+		file := entry.Files[idx]
 		log.VEventf(ctx, 2, "import file %s which starts at %s", file.Path, entry.Span.Key)
 
+		if err := rd.mem.Grow(ctx, bytesPerFile); err != nil {
```
This isn't the size of the file, but the overhead of the `ExternalSSTReader` that's reading the files. I'll rename the variable, but for now it's just a hard-coded guess based on benchmarks.

dt commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -400,6 +400,15 @@ func restore(
+			// After the change to bound the memory used by SST iterator, we can now
+			// get a progress entry for a partially completed span. Do not mark the
+			// span as done when receiving these entries.
+			if !progDetails.Done {
```
should sync up with @msbutler -- we were just talking about how this should maybe the a timestamp and then done := timestamp == endtime

rhu713 replied:
resolved and now sending the timestamp

dt commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -87,6 +95,16 @@ const restoreDataProcName = "restoreDataProcessor"
+// sstReaderOverheadBytesPerFile and sstReaderEncryptedOverheadBytesPerFile were obtained
+// benchmarking external SST iterators on GCP and AWS and selecting the highest
+// observed memory per file.
+const sstReaderOverheadBytesPerFile = 5 << 20
```
5MiB?! ooof, that's bigger than some of the actual files

rhu713 replied:
Yeah I was surprised as well.

dt commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -318,21 +368,74 @@ func (rd *restoreDataProcessor) openSSTs(
+			}
+
+			select {
+			case <-time.After(sstReaderMemWaitSec * time.Second):
```
can we do better than sleeping here?

What if we had a return channel or quotapool, so that when `cleanup` runs and shrinks, it unblocks the opener?

rhu713 replied:
Done, added a return channel.

dt commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -318,21 +381,83 @@ func (rd *restoreDataProcessor) openSSTs(
+			rd.mu.Unlock()
+
+			select {
+			case <-waitCh:
```
I'm trying to figure out if we might still have a lock order acquisition problem here:
What happens if right after we fail to `Grow()` all the pending flushes finish and release, before we set memWaitCh, so they breeze past it. then let's say our `sendIter` work also happen to run to completion, including  before we get around to setting `memWaitCh` below, so by the time we set it, _nobody_ is out there with a pending SST and going to call shrink and then close memWaitCh for us when it finishes?

dt replied:
I wonder if a single shared IntPool would be simpler here, instead of the mutex+ch? I haven't completely thought it through, so maybe not. In my head, it might look something like:  
* First, where we just `Grow()` today, we'd instead `TryAcquire` -- if that fails, we'd then try to `Grow()` and if that succeeds, update the pool capacity with the growth amount then go back to `TryAcquire`.
* If `TryAcquire` and `Grow()` both fail, we enter this block, send the iter, etc.
* Then we just do plain blocking Acquire() here, instead of mu/ch, and don't need the `continue` either: either via the TryAcquire or via this blocking Acquire, either way you reach the end of the block with an inc alloc reservation?

rhu713 replied:
done, now using a quota pool

dt commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -318,21 +368,88 @@ func (rd *restoreDataProcessor) openSSTs(
+						return err
+					}
+					storeFiles = storeFiles[:0]
+					iterAllocs = iterAllocs[:0]
```
I this slice resuse safe? aren't we handing these slices off in sendIter?

rhu713 replied:
🤦 oops, I'll just make new slices to keep everything simpler.

dt commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -343,10 +460,13 @@ func (rd *restoreDataProcessor) openSSTs(
+	ctx context.Context, ssts chan mergedSST, workerSSTChannels []chan mergedSST,
+) error {
 	return ctxgroup.GroupWorkers(ctx, rd.numWorkers, func(ctx context.Context, worker int) error {
 		kr, err := MakeKeyRewriterFromRekeys(rd.FlowCtx.Codec(), rd.spec.TableRekeys, rd.spec.TenantRekeys,
```
maybe here we should do ```if err := mem.Grow(minWorkerMemReservation); err != nil { if worker != 0 { return nil // no memory to run this worker} return errors.Wrap(err, "insufficient memory available to run restore") } else { // add minWorkerMemReservation to the pool``` to ensure the pool has a baseline amount (say 25MB) per worker ?

rhu713 replied:
done, reserved 25MB at the start of the worker.

lidorcarmel commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -343,11 +467,23 @@ func (rd *restoreDataProcessor) openSSTs(
+) error {
 	return ctxgroup.GroupWorkers(ctx, rd.numWorkers, func(ctx context.Context, worker int) error {
+		if err := rd.mem.Grow(ctx, minWorkerMemReservation); err != nil {
+			if worker != 0 {
+				return nil // no memory to run this worker
+			}
+			return errors.Wrap(err, "insufficient memory available to run restore")
```
what's the goal here? all workers except for the first one will silently exit, letting worker 0 continue running, but that worker 0 will not pick up work from the queues that other workers listen on, right?

dt replied:
Oh, I missed this: do we only pull work off of a specific queue? I imagined fewer workers would just mean draining a common queue slower and missed that we have per-worker queues outside the worker?

rhu713 replied:
Yeah I missed this too. I made the change so that the number of worker-specific queues match the final number of workers, PTAL.

rhu713 commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -318,21 +375,88 @@ func (rd *restoreDataProcessor) openSSTs(
+					}
+
+					log.VInfof(ctx, 2, "sending iterator after %d out of %d files due to insufficient memory", idx, len(entry.Files))
+					if err := sendIter(iter, dirs, workerSSTChannels[workerNum], iterAllocs, false); err != nil {
```
I have the workers pull off of their own channel after send iter fails to send an iterator for the entire span. I believe this is necessary since we don't want an out of order write between the first half of files and the second half of files for the same span, which is possible if all workers pull off of the same queue.

lidorcarmel replied:
I see why you did this, thanks, and thanks for the chat about this issue, it was helpful.

I think there are probably other ways to guarantee ordering, for example, what do you think about:
- moving the call to openSSTs into the worker itself (I was thinking of doing that anyway because we need to increase parallelism for openSSTs, either by having multiple goroutines or by moving it to the worker)
- then, each large request that needs to be fulfilled by multiple partial ssts can be done sequentially by a single worker, meaning, the worker will process the first partial sst, flush it, and then start processing the next one.

lidorcarmel commented:
I didn't review everything so sorry if it's written in the pr somewhere, but do you have estimations on how common it would be to write partial ssts? IIUC the first partial sst will be written to L6, the second to L5 and so on, and I assume if you have enough of those we will start writing to L0 (BTW it would be nice to write this somewhere if it's true, in some comment).

I have no idea whether this means that a tiny fraction will be written to L5 which will trigger a cheap compaction later, or that in some scenarios we would see a significant amount of bytes written to all layers and to L0 which will increase write amplification. I bet it's the former, but do we have a test or a back-of-the-envelop calculation or some estimation? maybe looking at pebble stats after a restore with many incremental layers?

if the answer is in the code please ignore me and I'll review early next week.

rhu713 commented:
Recording for posterity on the testing with 30MB of memory limit per restore data processor. Looks like most of the data from restore still ends up in L6. This may be due to the fact that the base layer always gets sent ingested first, thus guaranteeing that most of the data will go into L6.

```
__level_____count____size___score______in__ingest(sz_cnt)____move(sz_cnt)___write(sz_cnt)____read___r-amp___w-amp__val-bl
    WAL         1    22 M       -   3.2 G       -       -       -       -   3.2 G       -       -       -     1.0
      0         0     0 B    0.00   3.2 G   304 M   1.0 K     0 B       0   317 M     356     0 B       0     0.1     0 B
      1         0     0 B    0.00     0 B     0 B       0     0 B       0     0 B       0     0 B       0     0.0     0 B
      2         0     0 B    0.00     0 B     0 B       0     0 B       0     0 B       0     0 B       0     0.0     0 B
      3        23    62 M    0.98   323 M    16 M       4    15 M      11   1.3 G     409   1.3 G       1     4.2   1.6 M
      4        81   397 M    0.85   471 M   273 M      38    48 M      29   1.5 G     392   1.5 G       1     3.3   5.7 M
      5       235   1.6 G    0.47   387 M   1.2 G     152   8.1 M       6   602 M     171   605 M       1     1.6    17 M
      6      2985    25 G       -   298 M    25 G   3.2 K   1.6 K       1   425 M     168   437 M       1     1.4    26 M
  total      3324    27 G       -    30 G    27 G   4.4 K    71 M      47    34 G   1.5 K   3.9 G       4     1.1    50 M
  flush       208
compact       738     0 B     0 B       0          (size == estimated-debt, score = in-progress-bytes, in = num-in-progress)
  ctype       497       0     192      47       2       0       0  (default, delete, elision, move, read, rewrite, multi-level)
 memtbl         1    64 M
zmemtbl         0     0 B
   ztbl         0     0 B
 bcache     362 K    11 G   47.1%  (score == hit-rate)
 tcache       878   583 K   98.6%  (score == hit-rate)
  snaps         0       -       0  (score == earliest seq num)
 titers         0
 filter         -       -   85.3%  (score == utility)
 ```

rhu713 commented:
@lidorcarmel PTAL again. I've rebased this PR on top of your openSSTs to workers change. I've also added an env var `COCKROACH_RESTORE_MEM_FRACTION` in order to control the memory allocated to restore to be 0.5 of max SQL memory.

lidorcarmel commented on pkg/ccl/backupccl/restore_data_processor.go:
s/numWorkerSetting/maxRestoreWorkers/

rhu713 replied:
done

lidorcarmel reviewed (commented):
partial review :( sorry about that, I'll try to finish tonight or tomorrow.

lidorcarmel commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -296,18 +354,69 @@ func (rd *restoreDataProcessor) openSSTs(
+					return mSST, idx, err
+				}
+
+				alloc, err = rd.sstMemQuotaPool.Acquire(ctx, sstOverheadBytesPerFile)
```
so, the following can happen:
- thread 1 does `TryAcquire` and fails
- thread 1 `Grow`s the memory reservation
- thread 2 does `TryAcquire` and succeeds
- thread 1 get here to `Acquire` and hangs.
perhaps it's not a sever issue like a deadlock but if the memory monitor and the quota pool are abstracted away as a single object we can avoid those kind of issues, and for example in this case `Grow` and `TryAcquire` in a loop until we acquired or failed without hanging.

rhu713 replied:
added abstraction

lidorcarmel commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -246,14 +295,21 @@ func inputReader(
 
+// openSSTs opens all files in entry starting from the resumeIdx and returns a
+// multiplexed SST iterator over the files. If memory monitoring is enabled and
+// opening an additional  file would exceed the current memory budget, a partial
```
nit: double space.

lidorcarmel commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -343,22 +455,24 @@ func (rd *restoreDataProcessor) runRestoreWorkers(
-				if err != nil {
-					return done, err
+				idx := 0
+				for idx < len(entry.Files) {
```
I think it would be cleaner to return something like a `done` bool from openSSTs instead of comparing the idx to the num of files in the entry. the idx should be an opaque object that the caller should not use, just pass it back to openSSTs on the next call. (in that case we will probably want to rename it to something like restartFrom or maybe you'll have better ideas..)

rhu713 replied:
done

rhu713 commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -10936,3 +10936,121 @@ CREATE TABLE child_pk (k INT8 PRIMARY KEY REFERENCES parent);
+
+// Verify that restore with memory monitoring should be able to succeed with
+// partial SST iterators that shadow previously written values.
+func TestRestoreMemoryMonitoringWithShadowing(t *testing.T) {
```
The "tall" test was only there to verify that we have shadowing disabled and the restore is allowed to succeed without erroring with a collision error.

rhu713 replied:
I modified `TestRestoreMemoryMonitoring` to test for various number of splits in the base layer, number of incrementals, and memory limit on the processor.

rhu713 replied:
I modified `TestRestoreMemoryMonitoring` to test for various number of splits in the base layer, number of incrementals, and memory limit on the processor.

rhu713 commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -508,19 +623,31 @@ func (rd *restoreDataProcessor) processRestoreSpanEntry(
+	entry execinfrapb.RestoreSpanEntry,
+	pkIDs map[uint64]bool,
+	restoreTime hlc.Timestamp,
+	lastInSpan bool,
```
done, changed to timestamp

rhu713 commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -10936,3 +10936,121 @@ CREATE TABLE child_pk (k INT8 PRIMARY KEY REFERENCES parent);
+						}
+
+						restoreProcessorKnobCount.Add(1)
+					}}}}}
```
fixed

rhu713 commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -10936,3 +10936,121 @@ CREATE TABLE child_pk (k INT8 PRIMARY KEY REFERENCES parent);
 	}
 }
+
+// Verify that during restore, if an import span has too many files to fit in
```
changed to restore span

rhu713 commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -10936,3 +10936,121 @@ CREATE TABLE child_pk (k INT8 PRIMARY KEY REFERENCES parent);
+						// SST size, thus should all fit in one import span.
+						if len(entry.Files) != expectedNumFiles {
+							restoreProcessorKnobErr.Store(errors.Newf("expected entry to have %d files, got %d", expectedNumFiles, len(entry.Files)))
+						}
```
changed to require instead of the error

rhu713 commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -10936,3 +10936,121 @@ CREATE TABLE child_pk (k INT8 PRIMARY KEY REFERENCES parent);
+
+	sqlDB.Exec(t, "CREATE DATABASE data2")
+	sqlDB.Exec(t, "RESTORE data.bank FROM latest IN 'userfile:///backup' WITH OPTIONS (into_db='data2')")
+	require.Nil(t, restoreProcessorKnobErr.Load())
```
removed

rhu713 commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -10936,3 +10936,121 @@ CREATE TABLE child_pk (k INT8 PRIMARY KEY REFERENCES parent);
+
+	// Assert that the restore processor is processing the same span multiple
+	// times, and the count is based on what's expected from the memory budget.
+	require.Equal(t, 3, int(restoreProcessorKnobCount.Load())) // Ceiling(11/3)
```
fixed the denominator in the comment, it should still be ceiling

rhu713 commented on pkg/ccl/backupccl/restore_data_processor.go:
We need this error because the one way that this memory monitoring change can hang a job is if there's not enough memory to process a single file when we wait for quota. This gives the job some room for that.

lidorcarmel replied:
sorry I'm probably missing something - the behavior your mentioning is what we want, to always have at least a single worker that can do some work, but with that error we will fail without making progress, no? if all mem allocations fail, don't you want to succeed anyway with at least one worker?

rhu713 replied:
No, in this block I don't want the restore to continue if there's not enough memory to be reserved for one worker.

rhu713 commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -72,6 +76,10 @@ type restoreDataProcessor struct {
+
+	mon             *mon.BytesMonitor
+	mem             *mon.BoundAccount
+	sstMemQuotaPool *quotapool.IntPool
```
changed to using a single memory backed quota pool abstraction.

rhu713 commented on pkg/server/server_sql.go:
```diff
@@ -687,6 +687,9 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 	backupMemoryMonitor := execinfra.NewMonitor(ctx, bulkMemoryMonitor, "backup-mon")
+	restoreMemoryMonitor := mon.NewMonitorInheritWithLimit("restore-mon", bulk.RestoreMemLimit(cfg.MemoryPoolSize),
+		bulkMemoryMonitor)
+	restoreMemoryMonitor.StartNoReserved(ctx, bulkMemoryMonitor)
```
removed the extra monitor. Now the only new monitor is the one that's created as part of `MemoryBackedQuotaPool`

rhu713 commented on pkg/kv/bulk/setting.go:
```diff
@@ -24,6 +25,11 @@ var (
+	// restoreMemFraction is the maximum percentage of the SQL memory pool that
+	// could be used by restore. This can be overridden by environment variable.
+	restoreMemFraction = envutil.EnvOrDefaultFloat64("COCKROACH_RESTORE_MEM_FRACTION",
+		0.5)
```
changed to cluster setting per offline conversation

rhu713 commented on pkg/ccl/backupccl/restore_processor_planning.go:
```diff
@@ -111,13 +111,14 @@ func distRestore(
+			TenantRekeys:      dataToRestore.getTenantRekeys(),
+			PKIDs:             dataToRestore.getPKIDs(),
+			ValidateOnly:      dataToRestore.isValidateOnly(),
+			MemoryMonitorSSTs: true,
```
done, added cluster setting

rhu713 commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -296,18 +354,69 @@ func (rd *restoreDataProcessor) openSSTs(
 		storeFiles = append(storeFiles, storageccl.StoreFile{Store: dir, FilePath: file.Path})
-		// TODO(pbardea): When memory monitoring is added, return the currently
-		// accumulated iterators if we run into memory pressure.
+		idx++
```
done

rhu713 commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -296,18 +354,69 @@ func (rd *restoreDataProcessor) openSSTs(
+		file := entry.Files[idx]
 		log.VEventf(ctx, 2, "import file %s which starts at %s", file.Path, entry.Span.Key)
 
+		alloc, err := rd.sstMemQuotaPool.TryAcquire(ctx, sstOverheadBytesPerFile)
```
done, added abstraction

lidorcarmel commented on pkg/ccl/backupccl/restore_data_processor.go:
s/mas/max/ ?

rhu713 replied:
done

lidorcarmel commented on pkg/server/server_sql.go:
```diff
@@ -837,6 +837,7 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 		RangeStatsFetcher:        rangeStatsFetcher,
 		AdmissionPacerFactory:    cfg.admissionPacerFactory,
 		ExecutorConfig:           execCfg,
+		RootSQLMemoryPoolSize:    cfg.MemoryPoolSize,
```
is this a noop for sql? sorry I'm not sure I follow why this is needed..

rhu713 replied:
This should be. This allows the root SQL memory pool size to be available during the restore processor planning, since the amount of memory we want is limited to a fraction of that. I don't see anything else that exposes this value during restore, but maybe I missed something.

lidorcarmel commented on pkg/ccl/backupccl/backuputils/memory_backed_quota_pool.go:
```diff
@@ -0,0 +1,118 @@
+	quotaPool *quotapool.IntPool
+}
+
+// NewMemoryBackedQuotaPool creates an MemoryBackedQuotaPool from a
```
s/an/a/

rhu713 replied:
done

lidorcarmel commented on pkg/ccl/backupccl/backuputils/memory_backed_quota_pool.go:
```diff
@@ -0,0 +1,118 @@
+		//
+		// NB: we can update capacity despite concurrent usage of the quota pool
+		// because processing of requests in AbstractPool.Acquire is single threaded.
+		q.IncreaseCapacity(ctx, size)
```
I think you want to return a value based on the return value of IncreaseCapacity?
anyway, which brings me to the other point, can you add a `_test.go` file for this class?

rhu713 replied:
I added a test with a basic test case of a quota pool of varying sizes with varying number of users.

lidorcarmel reviewed (commented):
PTAL, can you also make sure you saw the older comments? I think David had some that were not addressed (or maybe fixed in the code but the comments have no response).

lidorcarmel commented on pkg/ccl/backupccl/backuputils/memory_backed_quota_pool_test.go:
```diff
@@ -0,0 +1,90 @@
+// goroutines acquiring and releasing quota, also with multiple
+// goroutines acquiring and releasing memory from the parent
+// memory monitor.
+func TestMemoryBackedQuotaPoolBasic(t *testing.T) {
```
this is a fancy test, I think you need a simple non-concurrent deterministic api unit tests, so that you can say allocate X and then verify it succeeded or failed, and of course verify more advanced scenarios.

a test like here may exercise more code path but also, it might be flaky and it will be hard to say what the issue is when it fails. also, nothing guarantees that the execution order of those goroutines will create a meaningful test - instead we may first run all the goroutines that grow and shrink the pool and then those that ask for quota. I understand it's not likely with 100 goroutines but overall my point is that a deterministic test can assert on the state of the object better. WDYT? did I miss anything about this test?

rhu713 commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -160,6 +220,20 @@ func (rd *restoreDataProcessor) Start(ctx context.Context) {
 	ctx = rd.StartInternal(ctx, restoreDataProcName)
 	rd.input.Start(ctx)
 
+	// First we reserve minWorkerMemReservation for each restore worker, and
+	// making sure that we always have enough memory for at least one worker. The
+	// maximum number of workers is based on the cluster setting. If the cluster
+	// setting is updated, the job should be PAUSEd and RESUMEd for the new
+	// setting to take effect.
+	numWorkers, err := reserveRestoreWorkerMemory(ctx, rd.flowCtx.Cfg.Settings, rd.qp)
+	if err != nil {
+		log.Warningf(ctx, "cannot reserve restore worker memory: %v", err)
+		rd.MoveToDraining(err)
+		return
+	}
```
Having the reserve memory happen before the processor does any work is a way to at least make sure the job fails fast if we suspect that there's not enough memory in the node to continue the restore. Restore is something that can take up a lot of cluster resources and I think failing the restore is before it starts is better than failing some heavy SQL query that's in progress. This pattern also exist in our backup job currently where we reserve the worker memory in the beginning, and fail the job if we are not able to do so.

We do have an issue here if the job replans at an unfortunate time and some other queries in the cluster prevents the job from resuming due to lack of memory, but I think in that case it's still better to fail the restore since it wasn't going to be able to continue anyways, and the user can always have an option to pause on error should this become a recurring problem.

lidorcarmel replied:
should this PR point to an issue? I'd like to understand the motivation for this change better. I can imagine 2 different motivations: one is that the current restore frequently causes issues such as OOMing nodes or causing queries to fail, and another is that everything is pretty much ok except for some rare failures but we want to increase the number of workers and therefore we need some memory limits. I guess I have the latter in my mind but maybe I'm wrong.
cc @dt to see what you think.

rhu713 commented on pkg/ccl/backupccl/bench_test.go:
```diff
@@ -147,3 +168,182 @@ func BenchmarkDatabaseFullBackup(b *testing.B) {
+			for _, encrypted := range []bool{true, false} {
+				for _, iterCount := range []int{1, 10, 100} {
+					for _, fileCount := range []int{10, 100, 1000} {
+						b.Run(fmt.Sprintf("fileCount=%d/iterCount=%d/rows=%d/store=%s/enc=%t", fileCount, iterCount, rows, testCase.storeType, encrypted), func(b *testing.B) {
```
fixed

rhu713 commented on pkg/ccl/backupccl/backuputils/memory_backed_quota_pool.go:
```diff
@@ -0,0 +1,127 @@
+// it to return the quota to the pool.
+// If the acquire fails because of not enough quota, it will repeatedly
+// attempt to increase the capacity of the pool until the acquire succeeds.
+// If  the capacity increase fails, then the function will return
```
fixed

rhu713 commented on pkg/ccl/backupccl/backuputils/memory_backed_quota_pool.go:
```diff
@@ -0,0 +1,127 @@
+// Returns true if the increase succeeds and false otherwise.
+//
+// Safe for concurrent use.
+func (q *MemoryBackedQuotaPool) IncreaseCapacity(ctx context.Context, size uint64) bool {
```
yeah I agree, returning err now.

rhu713 commented on pkg/ccl/backupccl/backuputils/memory_backed_quota_pool_test.go:
```diff
@@ -0,0 +1,90 @@
+package backuputils
+
+import (
+	"context"
```
I'm not sure, I don't see the 8 character tab in the editor now?

lidorcarmel reviewed (commented):
looks good except for that one issue.

related to that issue - can you explain the motivation for backporting this change?

rhu713 reviewed (commented):
The summary of offline thread is that 
1. we think that reserving worker memory for restore ahead of time (and failing the restore if we fail to reserve the minimum amount of memory is okay). and 
2. there will be a future PR to be smarter about the actual minimum number of workers 

w.r.t. the backport, that was initially added so it can go along with the 22.2 backport for slim manifests. I removed the backport label now since we've reverted the slim manifests backport as well.

lidorcarmel commented:
> 2. there will be a future PR to be smarter about the actual minimum number of workers

unfortunately I think we need a minimum in this pr, because without it we might run a restore with large machines using a single thread which will be a slow restore, and this would be a regression. I do agree that we don't need to be too smart about the min and max numbers. how about half of the threads is the min? meaning, if today we use N workers, with this PR we will only run restore of we can reserver `max(1, N/2)` workers. WDYT?

rhu713 commented:
> with this PR we will only run restore of we can reserve `max(1, N/2)` workers. WDYT?

Okay, I've changed the PR so that our min workers is `numRestoreWorkers/2` . Since by default `numRestoreWorkers` setting is `GOMAXPROCS - 1`, I think this should suffice as something that can be controlled by the number of cores but also something that can be overridden by a setting. 

I've also made a small change to reduce the minimum worker memory to `15MB`, which is still more than the memory required by 1 file. This is so that we can pass unit tests if the metamorphic value makes restore choose a minimum of 4 workers (by choosing `defaultNumWorkers=8`)

lidorcarmel reviewed (approved):
please fix the test failure, but other than that, lgtm!

rhu713 commented:
bors retry
